desc/objectives/_coils.py

```diff
@@ -0,0 +1,504 @@
+        values. If `normalize` is `True` and the target is in physical units,
+        this should also be set to True.
+        be set to True.
+    loss_function : {None, 'mean', 'min', 'max'}, optional
+        Loss function to apply to the objective values once computed. This loss function
+        is called on the raw compute value, before any shifting, scaling, or
+        normalization. Note: Has no effect for this objective.
+    deriv_mode : {"auto", "fwd", "rev"}
+        Specify how to compute jacobian matrix, either forward mode or reverse mode AD.
+        "auto" selects forward or reverse mode based on the size of the input and output
+        of the objective. Has no effect on self.grad or self.hess which always use
+        reverse mode and forward over reverse mode respectively.
+    grid : Grid, optional
+        Collocation grid containing the nodes to evaluate at.
+    name : str, optional
+        Name of the objective function.
+
+    """
+
+    def __init__(
+        self,
+        coil,
+        data_keys,
+        target=None,
+        bounds=None,
+        weight=1,
+        normalize=True,
+        normalize_target=True,
+        loss_function=None,
+        deriv_mode="auto",
+        grid=None,
+        name=None,
+    ):
+        self._grid = grid
+        self._data_keys = data_keys
+        self._normalize = normalize
+        super().__init__(
+            things=[coil],
+            target=target,
+            bounds=bounds,
+            weight=weight,
+            normalize=normalize,
+            normalize_target=normalize_target,
+            loss_function=loss_function,
+            deriv_mode=deriv_mode,
+            name=name,
+        )
+
+    def build(self, use_jit=True, verbose=1):
+        """Build constant arrays.
+
+        Parameters
+        ----------
+        use_jit : bool, optional
+            Whether to just-in-time compile the objective and derivatives.
+        verbose : int, optional
+            Level of output.
+
+        """
+        # local import to avoid circular import
+        from desc.coils import CoilSet, MixedCoilSet, _Coil
+
+        is_mixed_coils = isinstance(self.things[0], MixedCoilSet)
+        is_coil_set = isinstance(self.things[0], CoilSet)
+
+        coils = tree_flatten(
+            self.things[0],
+            is_leaf=lambda x: isinstance(x, _Coil) and not isinstance(x, CoilSet),
+        )[0]
+
+        # if using single coil, make coils and grid a list so they can be
+        # used with tree_map
+        coils = [coils[0]] if not is_coil_set else coils
+
+        if self._grid is None:
+            NFP = self.NFP if hasattr(self, "NFP") else 1
+            grid = lambda x: LinearGrid(N=2 * x + 5, NFP=NFP, endpoint=False)
+            self._grid = [grid(coil.N) for coil in coils]
+        elif self._grid.num_rho > 1 or self._grid.num_theta > 1:
+            raise TypeError("Only use toroidal resolution for coil grids.")
```

Comment:
this should be a `ValueError` since it's indicating that a value is wrong, a `TypeError` would be if they passed in something besides a grid like an array or int etc.